File: Arrays/slidingwindow/SlidingWindowExampleThree.java
package Arrays.slidingwindow;

/*
 * Example 3: 713. Subarray Product Less Than K.
 * 
 * Given an array of positive integers nums and an integer k,
 * return the number of subarrays where the product of all
 * the elements in the subarray is strictly less than k.
 * 
 * For example, given the input nums = [10, 5, 2, 6], k = 100,
 * the answer is 8. The subarrays with products less than k are:
 * 
 * [10], [5], [2], [6], [10, 5], [5, 2], [2, 6], [5, 2, 6]
 */

public class SlidingWindowExampleThree {

    public static void main(String[] args) {
        int[] nums = {10, 5, 2, 6};
        int k = 100;

        System.out.println(numSubarrayProductLessThanK(nums, k));
    }

    static int numSubarrayProductLessThanK(int[] nums, int k){

        if(k<= 1) {
            return 0;
        }

        int left = 0;
        int ans = 0;
        int curr = 1;
        
        for(int right = 0; right < nums.length; right++) {
            curr *= nums[right]; // collects product

            while(curr >= k) {
                curr /= nums[left];
                left++;
            }

            int subArrayFormula = right - left +1;

            ans += subArrayFormula; // be sure to add the values of the sub array
        }

        return ans;
    }

}

File: Arrays/slidingwindow/python/slidingwindowexamplethree.py
from typing import List

class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0
        
        curr = 1
        ans = 0
        left = 0

        for right in range(len(nums)):
            curr*=nums[right]

            while curr >= k:
                curr//=nums[left]
                left+=1
            
            ans += right-left+1
        
        return ans
    
    # Test the solution
if __name__ == "__main__":
    solution = Solution()
    
    # Test case 1
    nums = [10, 5, 2, 6]
    k = 100
    result = solution.numSubarrayProductLessThanK(nums, k)
    print(f"Input: nums = {nums}, k = {k}")
    print(f"Output: {result}")
    print(f"Expected: 8")
    print()
    
    # Test case 2
    nums = [10,5,2,6]
    k = 100
    result = solution.numSubarrayProductLessThanK(nums, k)
    print(f"Input: nums = {nums}, k = {k}")
    print(f"Output: {result}")
    print(f"Expected: 0")